lib/galaxy/webapps/galaxy/services/visualizations.py

```diff
@@ -60,3 +97,284 @@ def index(
+                if visualization.latest_revision
+                else None
+            ),
+            # need to encode ids in revisions as well
+            # NOTE: does not encode ids inside the configs
+            "revisions": [r.id for r in visualization.revisions],
+        }
+        # replace with trans.url_builder if possible
+        dictionary["url"] = url_for(
+            controller="visualization",
+            action="display_by_username_and_slug",
+            username=visualization.user.username,
+            slug=visualization.slug,
+        )
+        dictionary["username"] = visualization.user.username
+        dictionary["email_hash"] = md5_hash_str(visualization.user.email)
+        dictionary["tags"] = visualization.make_tag_string_list()
+        dictionary["annotation"] = get_item_annotation_str(trans.sa_session, trans.user, visualization)
+        app = cast(StructuredApp, trans.app)
+        if app.visualizations_registry:
+            visualizations_registry = cast(VisualizationsRegistry, app.visualizations_registry)
+            visualization_plugin = cast(VisualizationPlugin, visualizations_registry.get_plugin(dictionary["type"]))
+            dictionary["plugin"] = visualization_plugin.to_dict()
+        return VisualizationShowResponse(**dictionary)
+
+    def create(
+        self,
+        trans: ProvidesUserContext,
+        import_id: Optional[DecodedDatabaseIdField],
+        payload: VisualizationCreatePayload,
+    ) -> VisualizationCreateResponse:
+        """Returns a dictionary of the created visualization
+
+        :rtype:     dictionary
+        :returns:   dictionary containing Visualization details
+        """
+
+        if import_id:
+            visualization = self._import_visualization(trans, import_id)
+        else:
+            # must have a type (I've taken this to be the visualization name)
+            if not payload.type:
+                raise exceptions.RequestParameterMissingException("key/value 'type' is required")
+            type = payload.type
+            title = payload.title
+            slug = payload.slug
+            dbkey = payload.dbkey
+            annotation = payload.annotation
+            config = payload.config
+            save = payload.save
+
+            # generate defaults - this will err if given a weird key?
+            visualization = self._create_visualization(trans, type, title, dbkey, slug, annotation, save)
+
+            # Create and save first visualization revision
+            revision = VisualizationRevision(visualization=visualization, title=title, config=config, dbkey=dbkey)
+            visualization.latest_revision = revision
+
+            if save:
+                session = trans.sa_session
+                session.add(revision)
+                with transaction(session):
+                    session.commit()
+
+        return VisualizationCreateResponse(id=str(visualization.id))
```

Comment:
If you're not committing the visusalization visualization.id is None, in which case the VisualizationCreateResponse is invalid.